File: Task 2/Java/logistics/src/main/java/com/logistics/Components/Product/Plants.java
package com.logistics.Components.Product;

import java.util.Calendar;

public class Plants implements Flora {
    protected float cost;
    protected String name;
    protected String supplier;
    protected float discountAmount;
    protected Packaging packaging;
    protected boolean isCultivated;
    private boolean bearsFruit;

    public Plants() {
        cost = 0;
        name = "";
        supplier = "";
        discountAmount = 1;
        packaging = new Packaging(0, 0);
        isCultivated = false;
        bearsFruit = false;
    }

    public Plants(String name, float cost, String supplier, float discountAmount, int packingUnits, float costPerUnit, boolean isCultivated, boolean bearsFruit) {
        this.name = name;
        this.cost = cost;
        this.supplier = supplier;
        this.discountAmount = discountAmount;
        this.packaging = new Packaging(packingUnits, costPerUnit);
        this.isCultivated = isCultivated;
        this.bearsFruit = bearsFruit;
    }

    // Copy Constructor
    public Plants(Plants other) {
        this.name = other.name;
        this.cost = other.cost;
        this.supplier = other.supplier;
        this.discountAmount = other.discountAmount;
        this.packaging = new Packaging(other.packaging);
        this.isCultivated = other.isCultivated;
        this.bearsFruit = other.bearsFruit;
    }

    // Clone method
    @Override public Plants clone() {
        return new Plants(this);
    }

    // Getters
    @Override public float getCost() { return cost; }
    @Override public String getName() { return name; }
    @Override public String getSupplier() { return supplier; }
    @Override public float getDiscountAmount(){return discountAmount;}
    @Override public float getPackagingCost() { return packaging.calculateTotalCost(); }
    @Override public boolean getIsCultivated() { return isCultivated; }
    public boolean getBearsFruit() { return bearsFruit; }

    // Setters
    @Override public void setCost(float cost) { this.cost = cost; }
    @Override public void setName(String name) { this.name = name; }
    @Override public void setSupplier(String supplier) { this.supplier = supplier; }
    @Override public void setDiscountAmount(float discountAmount){this.discountAmount = discountAmount;}
    @Override public void setPackagingData(int packingUnits, float costPerUnit) { this.packaging = new Packaging(packingUnits, costPerUnit); }
    @Override public void setIsCultivated(boolean isCultivated) { this.isCultivated = isCultivated; }
    public void setBearsFruit(boolean bearsFruit) { this.bearsFruit = bearsFruit; }

    @Override public float calculateDiscount() {
        // Get the current month
        int currentMonth = Calendar.getInstance().get(Calendar.MONTH);

        // Check if the current month is December 
        if (currentMonth == Calendar.DECEMBER && (discountAmount >= 0 && discountAmount <= 1)) {
            return (1 - discountAmount);
        } else {
            return 1;
        }
    }

    @Override public String display_product_info() {
        return  "\nName: " + name +
                "\nCost: " + cost +
                "\nSupplier: " + supplier +
                "\nDiscount Amount: " + calculateDiscount() +
                "\nPackaging Cost: " + getPackagingCost() +
                "\nCultivated: " + isCultivated +
                "\nBears Fruit: " + bearsFruit;
    }
}

File: Task 3/C++Stock Viewer/Product/Plants.cpp
#include "Plants.h"
Plants::Plants() : Flora(){
    bearsFruit = false;
}
Plants::Plants(std::string name, float cost, std::string supplier, float discount_amount, int packing_units, float cost_per_unit, bool isCultivated, bool bearsFruit) 
        : Flora(name, cost, supplier, discount_amount, packing_units, cost_per_unit, isCultivated){
    this->bearsFruit = bearsFruit;
}

Plants::Plants(const Plants& other) : Flora(other) {
    this->bearsFruit = other.bearsFruit;
}

std::string Plants::display_product_info(){
    return Flora::display_product_info()+
            "\nBears Fruit: " +(bearsFruit? "true" : "false");
}

// ========================================== PROTOBUFFERS
void Plants::copyToProto(PBStockItem* message){
    PBPlants* pbPlants = message->mutable_plants();
    pbPlants->set_bearsfruit(bearsFruit);

    PBFlora* pbFlora = pbPlants->mutable_flora();
    pbFlora->set_iscultivated(isCultivated);

    PBProduct* pbProduct = pbFlora->mutable_product();
    pbProduct->set_name(name);
    pbProduct->set_cost(cost);
    pbProduct->set_supplier(supplier);

    PBPackage* pbPackage = pbProduct->mutable_packaging();
    pbPackage->set_units(packaging.getUnits());
    pbPackage->set_costperunit(packaging.getCost_Per_Unit());
}

void Plants::copyFromProto(const PBStockItem& message) {
    const PBPlants& pbPlants = message.plants();
    name = pbPlants.flora().product().name();
    cost = pbPlants.flora().product().cost();
    supplier = pbPlants.flora().product().supplier();
    packaging.setUnits(pbPlants.flora().product().packaging().units());
    packaging.setCost_Per_Unit(pbPlants.flora().product().packaging().costperunit());
    isCultivated = pbPlants.flora().iscultivated();
    bearsFruit = pbPlants.bearsfruit();
}